// ra-tree-ui-materialui/esm/TreeNodeActionsMenu.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) if (e.indexOf(p[i]) < 0)
            t[p[i]] = s[p[i]];
    return t;
};
import React, { Children, Component, isValidElement, cloneElement, } from 'react';
import IconButton from '@material-ui/core/IconButton';
import Menu from '@material-ui/core/Menu';
import MoreVertIcon from '@material-ui/icons/MoreVert';
var TreeNodeActionsMenu = /** @class */ (function (_super) {
    __extends(TreeNodeActionsMenu, _super);
    function TreeNodeActionsMenu() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.state = { anchorEl: null };
        _this.handleClick = function (event) {
            _this.setState({ anchorEl: event.currentTarget });
        };
        _this.handleClose = function () {
            _this.setState({ anchorEl: null });
        };
        return _this;
    }
    TreeNodeActionsMenu.prototype.render = function () {
        var _a = this.props, basePath = _a.basePath, children = _a.children, parentSource = _a.parentSource, positionSource = _a.positionSource, record = _a.record, resource = _a.resource, props = __rest(_a, ["basePath", "children", "parentSource", "positionSource", "record", "resource"]);
        var anchorEl = this.state.anchorEl;
        var open = Boolean(anchorEl);
        return (React.createElement(React.Fragment, null,
            React.createElement(IconButton, __assign({ "aria-label": "More", "aria-owns": open ? 'long-menu' : null, "aria-haspopup": "true", onClick: this.handleClick }, props),
                React.createElement(MoreVertIcon, null)),
            React.createElement(Menu, { id: "long-menu", anchorEl: anchorEl, open: open, onClose: this.handleClose, marginThreshold: 32 }, Children.map(children, function (child) {
                return isValidElement(child)
                    ? cloneElement(child, __assign({ basePath: basePath,
                        parentSource: parentSource,
                        positionSource: positionSource,
                        record: record,
                        resource: resource }, child.props))
                    : null;
            }))));
    };
    return TreeNodeActionsMenu;
}(Component));
export default TreeNodeActionsMenu;
