// ra-tree-ui-materialui/esm/Tree.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) if (e.indexOf(p[i]) < 0)
            t[p[i]] = s[p[i]];
    return t;
};
import React, { Children, Component, cloneElement, } from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import Card from '@material-ui/core/Card';
import { createStyles, withStyles } from '@material-ui/core/styles';
import { TreeController } from 'ra-tree-core';
import { Title } from 'ra-ui-materialui';
import withDragDropContext from './withDragDropContext';
import TreeListActions from './TreeListActions';
import TreeListToolbar from './TreeListToolbar';
var Tree = /** @class */ (function (_super) {
    __extends(Tree, _super);
    function Tree() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Tree.prototype.componentDidMount = function () {
        var childrenCount = Children.count(this.props.children);
        if (childrenCount > 1 && process.env.NODE_ENV !== 'production') {
            warnAboutChildren();
        }
    };
    Tree.prototype.render = function () {
        var _a = this.props, children = _a.children, parentSource = _a.parentSource, positionSource = _a.positionSource, props = __rest(_a, ["children", "parentSource", "positionSource"]);
        return (React.createElement(TreeController, __assign({ parentSource: parentSource, positionSource: positionSource }, props), function (controllerProps) { return (React.createElement(TreeView, __assign({}, controllerProps), children)); }));
    };
    Tree.propTypes = {
        actions: PropTypes.element,
        aside: PropTypes.element,
        basePath: PropTypes.string.isRequired,
        children: PropTypes.node,
        classes: PropTypes.object,
        className: PropTypes.string,
        exporter: PropTypes.func,
        filter: PropTypes.object,
        parentSource: PropTypes.string.isRequired,
        positionSource: PropTypes.string,
        resource: PropTypes.string.isRequired,
        title: PropTypes.string,
        version: PropTypes.number,
    };
    Tree.defaultProps = {
        classes: {},
        parentSource: 'parent_id',
        positionSource: 'position',
    };
    return Tree;
}(Component));
export { Tree };
export default withDragDropContext(Tree);
export var styles = createStyles({
    root: {
        display: 'flex',
        flex: 1,
    },
    card: {
        position: 'relative',
        flex: '1 1 auto',
    },
    actions: {
        zIndex: 2,
        display: 'flex',
        justifyContent: 'flex-end',
        flexWrap: 'wrap',
    },
});
var TreeView = withStyles(styles)(function (_a) {
    var _b = _a.actions, actions = _b === void 0 ? React.createElement(TreeListActions, null) : _b, aside = _a.aside, children = _a.children, className = _a.className, classes = _a.classes, defaultTitle = _a.defaultTitle, exporter = _a.exporter, filter = _a.filter, title = _a.title, version = _a.version, props = __rest(_a, ["actions", "aside", "children", "className", "classes", "defaultTitle", "exporter", "filter", "title", "version"]);
    return (React.createElement("div", __assign({ className: classnames('tree-page', classes.root, className) }, sanitizeRestProps(props)),
        React.createElement(Title, { title: title, defaultTitle: defaultTitle }),
        React.createElement(Card, { className: classes.card },
            actions && (React.createElement(TreeListToolbar, __assign({}, props, { actions: actions, exporter: exporter, permanentFilter: filter }))),
            React.createElement("div", { key: version }, cloneElement(children, props))),
        aside && cloneElement(aside, props)));
});
var warnAboutChildren = function () {
    return console.warn(
    // eslint-disable-line
    "You passed multiple children to the Tree component.\n\n    <Tree>\n        <TreeList>\n            <TreeNode>\n                <TextField source=\"name\" />\n            </TreeNode>\n        </TreeList>\n    </Tree>\n\nIf you need actions on each node, use the actions prop on the TreeNode component:\n\n    const MyNodeActions = props => (\n        <TreeNodeActions {...props}>\n            <EditButton />\n            <ShowButton />\n            <DeleteButton />\n        </TreeNodeActions>\n    );\n\n    <Tree>\n        <TreeList>\n            <TreeNode actions={<MyNodeActions />}>\n                <TextField source=\"name\" />\n            </TreeNode>\n        </TreeList>\n    </Tree>\n");
};
var sanitizeRestProps = function (_a) {
    var basePath = _a.basePath, children = _a.children, classes = _a.classes, closeNode = _a.closeNode, data = _a.data, expandNode = _a.expandNode, hasCreate = _a.hasCreate, hasEdit = _a.hasEdit, hasList = _a.hasList, hasShow = _a.hasShow, history = _a.history, loading = _a.loading, locale = _a.locale, location = _a.location, match = _a.match, nodes = _a.nodes, options = _a.options, parentSource = _a.parentSource, permissions = _a.permissions, positionSource = _a.positionSource, resource = _a.resource, toggleNode = _a.toggleNode, version = _a.version, rest = __rest(_a, ["basePath", "children", "classes", "closeNode", "data", "expandNode", "hasCreate", "hasEdit", "hasList", "hasShow", "history", "loading", "locale", "location", "match", "nodes", "options", "parentSource", "permissions", "positionSource", "resource", "toggleNode", "version"]);
    return rest;
};
