// ra-tree-ui-materialui/lib/TreeNodeActions.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var react_1 = __importStar(require("react"));
var styles_1 = require("@material-ui/core/styles");
var classnames_1 = __importDefault(require("classnames"));
var styles = function (theme) {
    return styles_1.createStyles({
        root: {
            marginLeft: 'auto',
            paddingRight: theme.spacing.unit * 3,
        },
    });
};
/**
 * Component used to position node actions correctly.
 *
 * // Usage with as a DropDown menu (recommanded)
 * @example
 * const TagNodeActions = props => (
 *     <TreeNodeActions {...props}>
 *         <TreeNodeActionsMenu {...props}>
 *             <AddChildNodeMenuItem />
 *             <AddNodeBeforeMenuItem />
 *             <AddNodeAfterMenuItem />
 *             <EditMenuItem />
 *             <DeleteMenuItem />
 *         </TreeNodeActionsMenu>
 *     </TreeNodeActions>
 * );
 * export const TagsList = (props) => (
 *     <Tree {...props}>
 *         <TreeList>
 *             <TreeNode actions={<TagNodeActions />}>
 *                 <TextField source="name" />
 *             </TreeNode>
 *         </TreeList>
 *     </Tree>
 * );
 * // Usage with standard react admin buttons
 * @example
 * const TagNodeActions = props => (
 *     <TreeNodeActions {...props}>
 *         <EditButton />
 *         <DeleteButton />
 *     </TreeNodeActions>
 * );
 * export const TagsList = (props) => (
 *     <Tree {...props}>
 *         <TreeList>
 *             <TreeNode actions={<TagNodeActions />}>
 *                 <TextField source="name" />
 *             </TreeNode>
 *         </TreeList>
 *     </Tree>
 * );
 */
var TreeNodeActions = styles_1.withStyles(styles)(function (_a) {
    var basePath = _a.basePath, children = _a.children, classes = _a.classes, className = _a.className, parentSource = _a.parentSource, positionSource = _a.positionSource, record = _a.record, resource = _a.resource;
    return (react_1.default.createElement("div", { className: classnames_1.default(classes.root, className) }, react_1.Children.map(children, function (child) {
        return react_1.isValidElement(child)
            ? react_1.cloneElement(child, __assign({ basePath: basePath,
                record: record,
                parentSource: parentSource,
                positionSource: positionSource,
                resource: resource }, child.props))
            : null;
    })));
});
exports.default = TreeNodeActions;
