// ra-tree-core/esm/actions.js
export var CRUD_GET_TREE_ROOT_NODES = 'RA/CRUD_GET_TREE_ROOT_NODES';
export var CRUD_GET_TREE_ROOT_NODES_SUCCESS = 'RA/CRUD_GET_TREE_ROOT_NODES_SUCCESS';
export var CRUD_GET_TREE_ROOT_NODES_LOADING = 'RA/CRUD_GET_TREE_ROOT_NODES_LOADING';
export var CRUD_GET_TREE_ROOT_NODES_FAILURE = 'RA/CRUD_GET_TREE_ROOT_NODES_FAILURE';
export var GET_TREE_ROOT_NODES = 'GET_TREE_ROOT_NODES';
export var crudGetTreeRootNodes = function (_a) {
    var resource = _a.resource, parentSource = _a.parentSource, positionSource = _a.positionSource;
    return ({
        type: CRUD_GET_TREE_ROOT_NODES,
        meta: {
            resource: resource,
            parentSource: parentSource,
            positionSource: positionSource,
            fetch: GET_TREE_ROOT_NODES,
        },
    });
};
export var CRUD_GET_TREE_CHILDREN_NODES = 'RA/CRUD_GET_TREE_CHILDREN_NODES';
export var CRUD_GET_TREE_CHILDREN_NODES_SUCCESS = 'RA/CRUD_GET_TREE_CHILDREN_NODES_SUCCESS';
export var CRUD_GET_TREE_CHILDREN_NODES_LOADING = 'RA/CRUD_GET_TREE_CHILDREN_NODES_LOADING';
export var CRUD_GET_TREE_CHILDREN_NODES_FAILURE = 'RA/CRUD_GET_TREE_CHILDREN_NODES_FAILURE';
export var GET_TREE_CHILDREN_NODES = 'GET_TREE_CHILDREN_NODES';
export var crudGetTreeChildrenNodes = function (_a) {
    var resource = _a.resource, parentSource = _a.parentSource, positionSource = _a.positionSource, nodeId = _a.nodeId;
    return ({
        type: CRUD_GET_TREE_CHILDREN_NODES,
        payload: { id: nodeId },
        meta: {
            resource: resource,
            parentSource: parentSource,
            positionSource: positionSource,
            fetch: GET_TREE_CHILDREN_NODES,
        },
    });
};
export var CRUD_MOVE_NODE = 'RA/TREE/CRUD_MOVE_NODE';
export var MOVE_NODE = 'MOVE_NODE';
export var crudMoveNode = function (_a) {
    var resource = _a.resource, data = _a.data, parentSource = _a.parentSource, positionSource = _a.positionSource, previousData = _a.previousData, _b = _a.refresh, refresh = _b === void 0 ? true : _b, redirectTo = _a.redirectTo, basePath = _a.basePath;
    return ({
        type: CRUD_MOVE_NODE,
        payload: {
            data: data,
            previousData: previousData,
        },
        meta: {
            resource: resource,
            parentSource: parentSource,
            positionSource: positionSource,
            fetch: MOVE_NODE,
            onSuccess: {
                notification: {
                    body: 'ra.notification.updated',
                    level: 'info',
                    messageArgs: {
                        smart_count: 1,
                    },
                },
                refresh: refresh,
                redirectTo: redirectTo,
                basePath: basePath,
            },
            onFailure: {
                notification: {
                    body: 'ra.notification.http_error',
                    level: 'warning',
                },
            },
        },
    });
};
export var TOGGLE_NODE = 'RA/TREE/TOGGLE_NODE';
export var toggleNode = function (_a) {
    var resource = _a.resource, nodeId = _a.nodeId;
    return ({
        type: TOGGLE_NODE,
        payload: nodeId,
        meta: { resource: resource },
    });
};
export var EXPAND_NODE = 'RA/TREE/EXPAND_NODE';
export var expandNode = function (_a) {
    var resource = _a.resource, nodeId = _a.nodeId;
    return ({
        type: EXPAND_NODE,
        payload: nodeId,
        meta: { resource: resource },
    });
};
export var CLOSE_NODE = 'RA/TREE/CLOSE_NODE';
export var closeNode = function (_a) {
    var resource = _a.resource, nodeId = _a.nodeId;
    return ({
        type: CLOSE_NODE,
        payload: nodeId,
        meta: { resource: resource },
    });
};
